File: 2015/python/01.py
def solve(inp, part):
    loc = 0

    for i,c in enumerate(inp):
        match c:
            case '(':
                loc += 1
            case ')':
                loc -= 1

        if part and loc < 0:
            return i + 1
    return loc

inp = open("../input/01").read()

print("Part 1:",solve(inp, False))
print("Part 2:",solve(inp, True))

File: 2015/rust/01.rs
fn main() {
    let input: Vec<char> = std::fs::read_to_string("../input/01").unwrap().chars().collect();

    println!("Part 1: {}", solve(&input, false));
    println!("Part 2: {}", solve(&input, true));
}

fn solve(lines: &[char], part: bool) -> i32 {
    let mut d: i32 = 0;

    for (i,c) in lines.iter().enumerate() {
        d += match c {
            '(' => 1,
            _ => -1
        };

        if part && d < 0 {
            return i as i32 + 1;
        }
    }

    d
}
